juice/src/net/container/sequential.rs

```diff
@@ -0,0 +1,317 @@
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+
+use crate::co::IBackend;
+use crate::net::{layer_from_config, Context, Descriptor, Inout, Layer, LayerConfig};
+use crate::util::LayerOps;
+
+// Config for a single child layer.
+#[derive(Debug, Clone, Default)]
+pub struct SequentialChildConfig {
+    name: String,
+    config: LayerConfig,
+    inputs: Vec<String>,
+    outputs: Vec<String>,
+}
+
+// Config for the sequential layer, listing child layers and (optionally) connections between them.
+#[derive(Debug, Clone, Default)]
+pub struct SequentialConfig {
+    // List of named inputs. If not given, only a single input is assumed.
+    inputs: Vec<String>,
+
+    // Sequence of layers.
+    layers: Vec<SequentialChildConfig>,
+
+    // List of named outputs. If not given, Sequential layer outputs will match the outputs
+    // of the last layer.
+    outputs: Vec<String>,
+}
+
+// A container layer that composes nested layers in a sequence.
+//
+// In the most simple case, all nested layers are executed one by one, with outputs of one layer
+// becoming inputs of the next one. Inputs of the Sequential layers are inputs to the first
+// nested layer, and outputs of the last nested layer are the outputs of the Sequential layer:
+//
+// ```
+// let mut cfg = SequentialConfig::new();
+// cfg.add_layer("linear", LayerConfig::Linear(LinearConfig{ output_size: 10}));
+// cfg.add_layer("softmax", LayerConfig::Softmax);
+// ```
+//
+// Sequential layer also supports complex flow graphs (as long as they are acyclic)
+// by allowing nested layers inputs and outputs to be mapped to named "buffers":
+//
+// ```
+// let mut cfg = SequentialConfig::new();
+// cfg.map_input("in");
+// cfg
+//   .add_layer("linear1", LayerConfig::Linear(LinearConfig{ output_size: 10}))
+//   .map_input("in")
+//   .map_output("linear1_out");
+// cfg
```

Comment:
I am not sure this is the best idea to enforce naming each layer. Passing Arcs seems an equally good idea?

Reply:
I am aware this is a common pattern, just thinking if the graph construction can be tightened. I am especially interested in anything that does dimensionality checks at compile time, but I have to cross check again the implications.

Reply:
> Passing Arcs seems an equally good idea?

Sorry, not sure what you mean. Maybe a code example?

> I am especially interested in anything that does dimensionality checks at compile time

Which checks you think we can implement? Layers dimensions are figured on the fly through the construction waterfall (next layer input dims match the previous layer output).

Reply:
My mental model was more in regards to make the dimensional match a compile time check. But I don't have a concrete impl idea yet nor am I sure it's a reasonable path or if that prevents some use-cases.

Reply:
And that idea put's up the need for `&'static str` as network scoped identifiers a bit, or at least would render them optional.
`Arc`s would resolve the issue that a layer would both be needed by the input as well as output layer.

Reply:
OK, no worries, we can always revisit later.